### PR TITLE
stop to strip 'impl' from impl trait type alias

### PR DESCRIPTION
**Pull request by rchaser53**
fix: #3815

### CODE REVIEW

File: tests/target/existential_type.rs [hunk 1]
```diff
@@ -6,10 +6,18 @@ pub type Adder<F, T>
 where
     T: Clone,
     F: Copy,
-= Fn(T) -> T;
```
Comment (rchaser53) [inline tests/target/existential_type.rs]: I guess this 'impl' should not be stripped. The compile result seems to be changed.
example is [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=27116dcfcb799c5bcba4e48f1af6a0a0)

File: tests/target/existential_type.rs [hunk 2]
```diff
@@ -6,10 +6,18 @@ pub type Adder<F, T>
...
+= impl Fn(T) -> T;
```
Comment (rchaser53) [inline tests/target/existential_type.rs]: I guess this 'impl' should not be stripped. The compile result seems to be changed.
example is [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=27116dcfcb799c5bcba4e48f1af6a0a0)

Comment (yelite) [reply tests/target/existential_type.rs]: Yes they are different things. Without `impl` it's equivalent to `dyn Fn(T) -> T`. 

Thanks for fixing it so quickly!

File: src/items.rs [hunk 3]
```diff
@@ -1528,7 +1528,7 @@ pub(crate) fn rewrite_opaque_type(
         context,
         indent,
         "type",
-        " =",
+        " = impl",
```
Comment (topecongiro) [inline src/items.rs]: This is not quite right, if the rhs consists of multiple lines then the `impl` will be orphaned:

```rust
pub type Type = impl
    VeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeryLongType<Target = i8>;
```

Comment (rchaser53) [reply src/items.rs]: Thank you the review. I fixed it.

File: src/expr.rs [hunk 4]
```diff
@@ -1857,14 +1857,14 @@ pub(crate) fn rewrite_assign_rhs<S: Into<String>, R: Rewrite>(
...
 
-pub(crate) fn rewrite_assign_rhs_with<S: Into<String>, R: Rewrite>(
+pub(crate) fn rewrite_assign_rhs_node<R: Rewrite>(
     context: &RewriteContext<'_>,
-    lhs: S,
+    lhs: &str,
     ex: &R,
     shape: Shape,
     rhs_tactics: RhsTactics,
+    has_impl: bool,
```
Comment (topecongiro) [inline src/expr.rs]: Hmm, adding a logic specific for handling `impl` to `rewrite_assign` does not seem a good idea to me.

Comment (rchaser53) [reply src/expr.rs]: I also think to use a wrapper GenericBounds and implement Rewrite trait is better than my idea. Thanks!

**Comment by rchaser53** [comment]
I added the commit to remove has_impl in GenericBounds. I don't think it necessary.

File: src/items.rs [hunk 5]
```diff
@@ -1118,6 +1118,18 @@ pub(crate) fn format_trait(
     }
 }
 
+struct GenericBounds<'a> {
```
Comment (topecongiro) [inline src/items.rs]: Nit: now that this type is specific to opaque types, I prefer to rename it to something like `OpaqueTypeBounds`.

Comment (rchaser53) [reply src/items.rs]: That's certainly true. I renamed it.

**Comment by topecongiro** [comment]
Thank you!

**Pull request merged**